### ISSUE TITLE
ML Drift GPU (Mali): DRQ int8 FULLY_CONNECTED + ADD on rank-3 input fails kernel creation — "Unable to parse bc coord for BATCH axis" (2.1.3 & 2.1.5)

### ISSUE THREAD

**Issue by john-rocky** [google-ai-edge/LiteRT#9277]
### Summary

On Mali (Pixel 8a), a dynamic-range-quantized (int8 channelwise weights, fp32 activations) `FULLY_CONNECTED` on a **rank-3** input whose output is consumed by an `ADD` fails GPU delegate kernel creation, and the whole `CompiledModel` compile fails with it:

```
E delegate_opencl.cc:330] Failed to create litert::ml_drift::DelegateKernelLiteRt:
  INVALID_ARGUMENT: Unable to parse bc coord for BATCH axis. Layout is bhwc,
  args are res, DST_X + 0, DST_Y + 0, DST_S + 0.
E tflite: Failed to initialize kernel.
```

Reproduced on **LiteRT 2.1.3 and 2.1.5** (`com.google.ai.edge.litert:litert`), identical error string (delegate_opencl.cc:328 on 2.1.3). Since `x + Linear(x)` is the residual pattern of every transformer/conformer block, this makes any DRQ-quantized transformer encoder fail GPU compile as a whole on Mali. The same graphs in f32 compile and run fine on the same device.

### Minimal repro (1 MB model, two ops)

```python
import torch
import litert_torch
from litert_torch.generative.quantize import quant_attrs, quant_recipe
from litert_torch.quantize import quant_config


class ResAdd(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = torch.nn.Linear(1024, 1024)

    def forward(self, x):
        return x + self.fc(x)


recipe = quant_recipe.LayerQuantRecipe(
    activation_dtype=quant_attrs.Dtype.FP32,
    weight_dtype=quant_attrs.Dtype.INT8,
    mode=quant_attrs.Mode.DYNAMIC_RANGE,
    algorithm=quant_attrs.Algorithm.MIN_MAX,
    granularity=quant_attrs.Granularity.CHANNELWISE,
)
qc = quant_config.QuantConfig(
    generative_recipe=quant_recipe.GenerativeQuantRecipe(default=recipe)
)

edge = litert_torch.signature("main", ResAdd().eval(), (torch.randn(1, 63, 1024),))
edge = edge.convert(enable_x64=False, quant_config=qc)
edge.export("drq_fc_add_repro.tflite")
```

(This is the same DRQ recipe the litert-samples `speech_recognition/convert` pipeline uses for its published i8 models.) Load on device with:

```kotlin
CompiledModel.create(path, CompiledModel.Options(Accelerator.GPU), null)  // -> "Failed to compile model"
```

### What compiles vs. what does not (all same device, same recipe)

| graph | GPU compile |
| --- | --- |
| DRQ `Linear` alone — rank-2, rank-3, or rank-4 input | ✅ OK |
| DRQ `x + Linear(x)`, rank-2 input `[1, 1024]` | ✅ OK |
| **DRQ `x + Linear(x)`, rank-3 input `[1, 63, 1024]`** | ❌ fails as above |
| DRQ `Linear(x) + y` (two graph inputs), rank-3 | ❌ fails as above |
| the identical rank-3 graph, f32 weights | ✅ OK |
| DRQ Conv2D / LSTM / Embedding single-op probes | ✅ OK |

So the isolated op is fine; the failure appears when the int8-weight FC's output feeds an `ADD` in the batched (rank-3, bhwc) layout — it looks like the fused FC+ADD epilogue's batch-coordinate codegen.

### Environment

- Device: Pixel 8a (Tensor G3, Mali-G715), Android 16
- Runtime: `com.google.ai.edge.litert:litert` 2.1.5 and 2.1.3, `CompiledModel` + `Accelerator.GPU` (OpenCL / ML Drift)
- Converter: litert-torch 0.10.0, recipe above

### Where it was found

`litert-community/parakeet-tdt_ctc-0.6b-ja` — the i8 DRQ file fails whole-model GPU compile on the Pixel 8a with exactly this signature (the f32 file compiles with every node delegated: 1862/1862 + 2083/2083). The same DRQ layout in the litert-community Adreno/Tensor-targeted uploads is presumably fine on those SoCs; this looks Mali-specific.

I've kept the generated repro file and the probe harness, so I'm happy to re-run this against candidate builds or gather more logs if useful.

**Comment by shivvaam0001** [comment]
Hi @john-rocky,

Thank you for reaching out!
Have you been able to test this issue with LiteRT 2.1.6 or a newer version?

Thanks!

**Comment by john-rocky** [comment]
Hi @shivvaam0001, thanks for following up. Yes — re-tested today on 2.2.0 (the newest on Maven, released yesterday): the failure is unchanged. Same Pixel 8a, same minimal repro from the issue body (`drq_fc_add_repro.tflite`), Kotlin CompiledModel + `Accelerator.GPU`:

```
E delegate_opencl.cc:340] Failed to create litert::ml_drift::DelegateKernelLiteRt:
  INVALID_ARGUMENT: Unable to parse bc coord for BATCH axis. Layout is bhwc,
  args are res, DST_X + 0, DST_Y + 0, DST_S + 0.; selector object=dst_tensor,
  selector=Write, args=[res, DST_X + 0, DST_Y + 0, DST_S + 0], template_args=[],
  descriptor=float16, TensorStorageType::TEXTURE_2D, layout: bhwc, shape: {bhwc, {63, 1, 1, 1024}}
E tflite: Failed to initialize kernel.
```

Same error string as 2.1.3 (delegate_opencl.cc:328) and 2.1.5 (:330). One clue that is new in 2.2.0's logging: immediately before the failure, the model builder reports a batch-dimension mismatch on the fused FC's output —

```
W gpu_model_builder.cc:3837] Shape mismatch in GpuModelBuilder::UpdateOutputTensors(output of
  convolution1x1(conv_generic)). Requested shape:{bhwdc, {63, 1, 1, 1, 1024}} vs reference
  shape: {bhwdc, {1, 1, 63, 1, 1024}}
```

— the rank-3 middle dim (63) sits in the batch slot in one layout and in W in the other, which fits the downstream "bc coord for BATCH axis" parse failure. Happy to re-run against any candidate build.

**Comment by aravindmurali22** [comment]
Hi @fengwuyao , following up on this thread, is there any recent updates on this? Thanks!
